eth/stagedsync/stage_bodies.go

```diff
@@ -252,15 +344,35 @@ Loop:
 			log.Trace("bodyLoop woken up by the incoming request")
 		}
 		d6 += time.Since(start)
+
+		return false, nil
 	}
-	if err := s.Update(tx, bodyProgress); err != nil {
-		return err
+
+	// kick off the loop and check for any reason to stop and break early
+	for !stopped {
+		shouldBreak, err := loopBody()
+		if err != nil {
+			return err
+		}
+		if shouldBreak {
+			break
+		}
 	}
+
+	// remove the temporary bucket for bodies stage
 	if !useExternalTx {
-		if err := tx.Commit(); err != nil {
```

Comment:
seems this line lost

Reply:
This is replaced with the check on line 299, if work has been done in the current loop it will be committed